semperfiwebdesign/all-in-one-seo-pack
# Refactor AIOSEOP_Notices notice operations to include hook

Issue #2516

## Proposed changes

Conceptually, this changes individual notice load operations to use a hook for accessing notice config data, rather than storing everything in the database. Just changing to a hook has a number of implications.

As mentioned in issue.

* Use a hook to load notice config data.
    * Which would allow extended notices from other plugins (Pro).
* Remove duplicate code.
* Reduce the Database/SQL.
* Auto-load notices of their own individual files.
    * Should also reduce merge conflicts.
* Prevent having to always update the database to reflect the notice data changes.
    * Currently, updating the notice message would require a database update.

### Hook Added

This is mostly for internal purposes, but it's possible other plugins could use the API (including AIOSEOP Pro).

**aioseop_admin_notice-{$notice_slug}**

`add_filter( 'aioseop_admin_notice-{$notice_slug}', 'notice_function_name', 10, 1 );`

#### Parameters

**$notice_data**
(array) Contains the notice config data.
Default: array()

## Types of changes

What types of changes does your code introduce?
_Delete those that don't apply_

- Bugfix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Adds new API hooks
- Removing old code/functionality
- Improves existing functionality
- Improves existing code

## Checklist


## Testing instructions

1. Activate the plugin, and check to see if notices are still operating.

Some of these notices may need to have their delay time reduced; 'review_plugin' is one with a long delay time.

**Blog Visibility**

1. Go to Settings > Reading.
2. Check/Enable 'Discourage search engines from indexing this site'.

**WooCommerce**

1. Install/Activate WooCommerce.

**Sitemap Indexes**

1. Have over 1000 posts in order to trigger.
2. Disable indexes in Sitemap settings.

**Review Plugin**

1. Just have plugin active.

### Resetting Notices

You can also reset notices by doing the following...

```
function aioseop_reset_notice() {
	global $aioseop_notices;
	$aioseop_notices->reset_notice( 'NOTICE_SLUG' );
}
add_action( 'admin_init', 'aioseop_reset_notice' );
```

#### SQL Method

Deleting notices in database

1. Go to localhost > phpMyAdmin, and go to target database table.
2. Go to wp_options.
3. Click Search tab, and in 'option_name' type `%aioseop%.
4. Delete `aioseop_notices`.

OR for user notices

1. Go to localhost > phpMyAdmin, and go to target database table.
2. Go to wp_usermeta.
3. Click Search tab, and in 'meta_key' type `%aioseop%.
4. Delete `aioseop_notice_*`.

## Further Comments

This will still need the unit tests updated, but this is ready for testing.
